File: ros2_foxy_tutorial_ws/src/my_cpp_pkg_/src/my_first_node.cpp
#include "rclcpp/rclcpp.hpp"

class MyNode : public rclcpp::Node
{
public:
    // create the node, give the name of the node
    MyNode() : Node("cpp_test"), counter_(0){
        RCLCPP_INFO(this->get_logger(), "Hello Cpp Node");

        timer_ = this->create_wall_timer(std::chrono::seconds(1),
                                         std::bind(&MyNode::timerCallback, this));
    }

private:
    void timerCallback(){
        counter_++;
        RCLCPP_INFO(this->get_logger(), "Hello %d", counter_);
    }

    rclcpp::TimerBase::SharedPtr timer_;
    int counter_;
};

int main(int argc, char **argv){
    rclcpp::init(argc, argv);

    auto node = std::make_shared<MyNode>();
    // pause the program here
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_foxy_tutorial_ws/src/my_py_pkg/my_py_pkg/my_first_node.py
#!/usr/bin/env python3

# to use ros2 functionalities
import rclpy
from rclpy.node import Node

class MyNode(Node):

    def __init__(self):
        super().__init__("py_test") # name of the node
        self.counter_ = 0
        self.get_logger().info("Hello ROS2!!!")
        # period between two callbacks, 0.5 seconds -> 2 Hertz
        self.create_timer(0.5, self.timer_callback)

    def timer_callback(self):
        self.counter_ += 1
        self.get_logger().info("Hello " + str(self.counter_))

def main(args=None):
    # initialize ros2 communication
    rclpy.init(args=args) # pass arguments
    node = MyNode()
    # pause the program here, allow your node to be alive
    # later callbacks will be able to be called from spin()
    rclpy.spin(node) 
    rclpy.shutdown() # last line

if __name__ == "__main__":
    main()